File: android/app/src/main/java/com/magicforms/mobile/ui/AuthViewModel.kt
package com.magicforms.mobile.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.magicforms.mobile.data.OfflineSubmitQueue
import com.magicforms.mobile.data.PushTokenStore
import com.magicforms.mobile.data.SessionExpiredException
import com.magicforms.mobile.data.TokenStore
import com.magicforms.mobile.data.api.AuthApi
import com.magicforms.mobile.data.api.DeviceApi
import com.magicforms.mobile.data.api.HomeApi
import android.provider.Settings
import com.magicforms.mobile.data.models.ApiEntity
import com.magicforms.mobile.data.models.ApiException
import com.magicforms.mobile.data.models.ApiUser
import com.magicforms.mobile.data.models.HomeSummaryResponse
import com.magicforms.mobile.data.models.InboxItem
import com.magicforms.mobile.data.models.PublishedFormItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

data class AuthUiState(
    val entitySlug: String = "",
    val username: String = "",
    val password: String = "",
    val directoryEntities: List<ApiEntity> = emptyList(),
    val user: ApiUser? = null,
    val entities: List<ApiEntity> = emptyList(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null,
    val token: String? = null,
    val homeSummary: HomeSummaryResponse? = null,
    val publishedForms: List<PublishedFormItem> = emptyList(),
    val inboxItems: List<InboxItem> = emptyList(),
    val inboxTotal: Int = 0,
    val isLoadingHome: Boolean = false,
    val inboxActingIds: Set<Int> = emptySet(),
    val inboxItemErrors: Map<Int, String> = emptyMap(),
    /** False until a saved token (if any) has been validated — avoids flashing the login screen. */
    val isSessionReady: Boolean = false,
    val sessionExpiredMessage: String? = null,
    val pendingSyncCount: Int = 0,
) {
    val isLoggedIn: Boolean get() = token != null && user != null

    val requiresEntitySlug: Boolean get() = directoryEntities.isNotEmpty()

    val inboxBadgeCount: Int
        get() = maxOf(homeSummary?.inboxCount ?: 0, inboxTotal)

    val bannerEntityLabel: String
        get() {
            val fromSummary = homeSummary?.bannerEntityName?.trim().orEmpty()
            if (fromSummary.isNotBlank()) return fromSummary
            return entities.map { it.name }.filter { it.isNotBlank() }.joinToString(" · ")
        }

    val bannerLogoUrl: String?
        get() {
            homeSummary?.bannerLogoUrl?.trim()?.takeIf { it.isNotBlank() }?.let { return it }
            if (entities.size == 1) {
                return entities[0].logoUrl?.trim()?.takeIf { it.isNotBlank() }
            }
            return null
        }

    val canSubmitLogin: Boolean
        get() {
            val userOk = username.trim().isNotEmpty()
            val entityOk = !requiresEntitySlug || entitySlug.trim().isNotEmpty()
            return userOk && entityOk && password.isNotEmpty() && !isLoading
        }
}

class AuthViewModel(application: Application) : AndroidViewModel(application) {
    private val authApi = AuthApi()
    private val homeApi = HomeApi()
    private val deviceApi = DeviceApi()
    private val tokenStore = TokenStore(application)
    private val offlineQueue = OfflineSubmitQueue(application)

    private val _state = MutableStateFlow(AuthUiState())
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    init {
        bootstrap()
    }

    fun onEntitySlugChange(value: String) = _state.update { it.copy(entitySlug = value) }
    fun onUsernameChange(value: String) = _state.update { it.copy(username = value) }
    fun onPasswordChange(value: String) = _state.update { it.copy(password = value) }

    private fun bootstrap() {
        viewModelScope.launch {
            try {
                tokenStore.load()?.let { saved ->
                    _state.update { it.copy(token = saved) }
                    validateStoredSession()
                }
                loadDirectoryEntities()
            } finally {
                _state.update { it.copy(isSessionReady = true) }
            }
        }
    }

    private suspend fun validateStoredSession(): Boolean {
        val token = _state.value.token ?: return false
        return try {
            val response = withContext(Dispatchers.IO) { authApi.me(token) }
            _state.update {
                it.copy(
                    user = response.user,
                    entities = response.entities.orEmpty(),
                    errorMessage = null,
                )
            }
            loadHomeDataAwait()
            registerPushTokenIfAvailable()
            flushOfflineQueue()
            true
        } catch (_: SessionExpiredException) {
            handleSessionExpired()
            false
        } catch (_: Exception) {
            signOutLocal()
            false
        }
    }

    fun handleSessionExpired() {
        if (_state.value.token == null) return
        _state.update {
            it.copy(sessionExpiredMessage = "Your session expired. Please sign in again.")
        }
        signOutLocal()
    }

    fun clearSessionExpiredMessage() {
        _state.update { it.copy(sessionExpiredMessage = null) }
    }

    fun flushOfflineQueue() {
        val token = _state.value.token ?: return
        viewModelScope.launch {
            val synced = withContext(Dispatchers.IO) { offlineQueue.flush(token) }
            _state.update { it.copy(pendingSyncCount = offlineQueue.pendingCount()) }
            if (synced > 0) loadHomeData()
        }
    }

    fun refreshPendingSyncCount() {
        _state.update { it.copy(pendingSyncCount = offlineQueue.pendingCount()) }
    }

    fun offlineQueue(): OfflineSubmitQueue = offlineQueue

    private suspend fun registerPushTokenIfAvailable() {
        val bearer = _state.value.token ?: return
        val push = PushTokenStore.load(getApplication()) ?: return
        val deviceId = Settings.Secure.getString(
            getApplication().contentResolver,
            Settings.Secure.ANDROID_ID,
        ).orEmpty()
        runCatching {
            withContext(Dispatchers.IO) {
                deviceApi.registerPushToken(bearer, push, deviceId)
            }
        }
    }

    fun loadDirectoryEntities() {
        viewModelScope.launch {
            runCatching {
                withContext(Dispatchers.IO) { authApi.fetchDirectoryEntities() }
            }.onSuccess { list ->
                _state.update { it.copy(directoryEntities = list) }
            }
        }
    }

    fun login() {
        val current = _state.value
        val trimmedUser = current.username.trim()
        val trimmedSlug = current.entitySlug.trim()

        if (trimmedUser.isEmpty()) {
            _state.update { it.copy(errorMessage = "Enter your username.") }
            return
        }
        if (current.requiresEntitySlug && trimmedSlug.isEmpty()) {
            _state.update { it.copy(errorMessage = "Enter your organization slug (e.g. mosa-kuwait).") }
            return
        }

        val slugForApi = trimmedSlug.ifBlank { null }
        _state.update { it.copy(isLoading = true, errorMessage = null) }

        viewModelScope.launch {
            try {
                val response = withContext(Dispatchers.IO) {
                    authApi.login(trimmedUser, current.password, slugForApi)
                }
                val tokenValue = response.token
                    ?: run {
                        _state.update { it.copy(isLoading = false, errorMessage = "No token returned.") }
                        return@launch
                    }
                tokenStore.save(tokenValue)
                _state.update {
                    it.copy(
                        isLoading = false,
                        token = tokenValue,
                        user = response.user,
                        entities = response.entities.orEmpty(),
                        password = "",
                        errorMessage = null,
                    )
                }
                registerPushTokenIfAvailable()
                flushOfflineQueue()
                loadHomeData()
            } catch (e: ApiException) {
                val msg = if (e.code == "organization_required") {
                    e.entities?.takeIf { it.isNotEmpty() }?.let { list ->
                        _state.update { s -> s.copy(directoryEntities = list) }
                    }
                    "Enter your organization slug, then your username."
                } else {
                    e.message
                }
                _state.update { it.copy(isLoading = false, errorMessage = msg) }
            } catch (e: Exception) {
                _state.update {
                    it.copy(isLoading = false, errorMessage = e.message ?: "Sign-in failed.")
                }
            }
        }
    }

    fun refreshSession() {
        val token = _state.value.token ?: return
        _state.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            val ok = validateStoredSession()
            _state.update {
                it.copy(
                    isLoading = false,
                    errorMessage = if (ok) null else "Session expired. Please sign in again.",
                )
            }
        }
    }

    fun loadHomeData() {
        val token = _state.value.token ?: return
        _state.update { it.copy(isLoadingHome = true) }
        viewModelScope.launch {
            try {
                loadHomeDataAwait()
            } catch (_: SessionExpiredException) {
                handleSessionExpired()
            }
        }
    }

    private suspend fun loadHomeDataAwait() {
        val token = _state.value.token ?: return
        _state.update { it.copy(isLoadingHome = true) }
        var summary: HomeSummaryResponse? = null
        var forms: List<PublishedFormItem>? = null
        var inboxItems: List<InboxItem> = emptyList()
        var inboxTotal = 0
        var error: String? = null

        withContext(Dispatchers.IO) {
            runCatching { homeApi.fetchSummary(token) }
                .onSuccess { summary = it }
                .onFailure { e ->
                    if (e is SessionExpiredException) throw e
                    error = e.message ?: "Failed to load summary."
                }

            runCatching { homeApi.fetchPublishedForms(token).forms }
                .onSuccess { forms = it }
                .onFailure {
                    if (error == null) error = it.message ?: "Failed to load forms."
                }

            runCatching { homeApi.fetchInbox(token) }
                .onSuccess {
                    inboxItems = it.items
                    inboxTotal = it.total
                }
                .onFailure {
                    if (error == null) error = it.message ?: "Failed to load inbox."
                }
        }

        _state.update {
            it.copy(
                isLoadingHome = false,
                homeSummary = summary ?: it.homeSummary,
                publishedForms = forms ?: it.publishedForms,
                inboxItems = inboxItems,
                inboxTotal = inboxTotal,
                errorMessage = error,
                pendingSyncCount = offlineQueue.pendingCount(),
            )
        }
    }

    fun refreshInbox() {
        val token = _state.value.token ?: return
        _state.update { it.copy(isLoadingHome = true) }
        viewModelScope.launch {
            try {
                val result = withContext(Dispatchers.IO) {
                    coroutineScope {
                        val summary = async { homeApi.fetchSummary(token) }
                        val inbox = async { homeApi.fetchInbox(token) }
                        Pair(summary.await(), inbox.await())
                    }
                }
                _state.update {
                    it.copy(
                        isLoadingHome = false,
                        homeSummary = result.first,
                        inboxItems = result.second.items,
                        inboxTotal = result.second.total,
                    )
                }
            } catch (e: Exception) {
                _state.update {
                    it.copy(isLoadingHome = false, errorMessage = e.message ?: "Failed to refresh inbox.")
                }
            }
        }
    }

    fun performInboxWorkflow(
        submissionId: Int,
        decision: String,
        comment: String?,
        workflowActionAnchor: Int?,
        onSuccess: () -> Unit = {},
    ) {
        val token = _state.value.token ?: return
        _state.update {
            it.copy(
                inboxActingIds = it.inboxActingIds + submissionId,
                inboxItemErrors = it.inboxItemErrors - submissionId,
            )
        }
        viewModelScope.launch {
            try {
                withContext(Dispatchers.IO) {
                    homeApi.performWorkflow(token, submissionId, decision, comment, workflowActionAnchor)
                    val summary = homeApi.fetchSummary(token)
                    val inbox = homeApi.fetchInbox(token)
                    summary to inbox
                }.let { (summary, inbox) ->
                    _state.update {
                        it.copy(
                            homeSummary = summary,
                            inboxItems = inbox.items,
                            inboxTotal = inbox.total,
                        )
                    }
                }
                onSuccess()
            } catch (e: ApiException) {
                _state.update {
                    it.copy(inboxItemErrors = it.inboxItemErrors + (submissionId to (e.message ?: "Workflow failed.")))
                }
            } catch (e: Exception) {
                _state.update {
                    it.copy(inboxItemErrors = it.inboxItemErrors + (submissionId to (e.message ?: "Workflow failed.")))
                }
            } finally {
                _state.update { it.copy(inboxActingIds = it.inboxActingIds - submissionId) }
            }
        }
    }

    fun logout() {
        val token = _state.value.token
        val push = PushTokenStore.load(getApplication())
        _state.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            if (token != null) {
                runCatching {
                    withContext(Dispatchers.IO) {
                        if (push != null) deviceApi.unregisterPushToken(token, push)
                        authApi.logout(token)
                    }
                }
            }
            signOutLocal()
            _state.update { it.copy(isLoading = false) }
        }
    }

    private fun signOutLocal() {
        tokenStore.delete()
        _state.update {
            it.copy(
                token = null,
                user = null,
                entities = emptyList(),
                password = "",
                homeSummary = null,
                publishedForms = emptyList(),
                inboxItems = emptyList(),
                inboxTotal = 0,
                inboxActingIds = emptySet(),
                inboxItemErrors = emptyMap(),
                pendingSyncCount = offlineQueue.pendingCount(),
            )
        }
    }
}

File: ios/MagicFormsMobile/ViewModels/AuthViewModel.swift
import Foundation

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var entitySlug = ""
    @Published var username = ""
    @Published var password = ""
    @Published private(set) var directoryEntities: [APIEntity] = []
    @Published var user: APIUser?
    @Published var entities: [APIEntity] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    @Published private(set) var homeSummary: HomeSummary?
    @Published private(set) var publishedForms: [PublishedFormItem] = []
    @Published private(set) var inboxItems: [InboxItem] = []
    @Published private(set) var inboxTotal = 0
    @Published private(set) var isLoadingHome = false
    @Published private(set) var inboxActingIds: Set<Int> = []
    @Published var inboxItemErrors: [Int: String] = [:]
    /// False until keychain token (if any) has been validated — avoids flashing the login screen.
    @Published private(set) var isSessionReady = false
    @Published var sessionExpiredMessage: String?

    private(set) var token: String?
    private var sessionObserver: NSObjectProtocol?

    /// Studio nav badge; fall back to loaded inbox list if summary is not ready yet.
    var inboxBadgeCount: Int {
        let fromSummary = homeSummary?.inboxCount ?? 0
        return max(fromSummary, inboxTotal)
    }

    /// Organization line for the app bar (same as studio header).
    var bannerEntityLabel: String {
        let fromSummary = homeSummary?.bannerEntityName.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !fromSummary.isEmpty { return fromSummary }
        return entities.map(\.name).filter { !$0.isEmpty }.joined(separator: " · ")
    }

    var bannerLogoURL: URL? {
        if let url = homeSummary?.bannerLogoUrl.trimmingCharacters(in: .whitespacesAndNewlines),
           !url.isEmpty,
           let parsed = URL(string: url) {
            return parsed
        }
        if entities.count == 1,
           let url = entities[0].logoUrl?.trimmingCharacters(in: .whitespacesAndNewlines),
           !url.isEmpty,
           let parsed = URL(string: url) {
            return parsed
        }
        return nil
    }

    var isLoggedIn: Bool { token != nil && user != nil }

    /// Directory sign-in needs an organization slug whenever any directory org is configured.
    var requiresEntitySlug: Bool {
        !directoryEntities.isEmpty
    }

    var canSubmitLogin: Bool {
        let userOk = !username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let entityOk = !requiresEntitySlug
            || !entitySlug.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        return userOk && entityOk && !password.isEmpty
    }

    init() {
        sessionObserver = NotificationCenter.default.addObserver(
            forName: .mobileSessionExpired,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in
                self?.handleSessionExpired()
            }
        }
    }

    deinit {
        if let sessionObserver {
            NotificationCenter.default.removeObserver(sessionObserver)
        }
    }

    func bootstrap() async {
        if let saved = KeychainTokenStore.load() {
            token = saved
            await validateStoredSession()
        }
        await loadDirectoryEntities()
        isSessionReady = true
    }

    func loadDirectoryEntities() async {
        do {
            directoryEntities = try await AuthAPI.fetchDirectoryEntities()
        } catch {
            directoryEntities = []
        }
    }

    func login() async {
        errorMessage = nil
        isLoading = true
        defer { isLoading = false }

        let trimmedUser = username.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedSlug = entitySlug.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmedUser.isEmpty {
            errorMessage = "Enter your username."
            return
        }

        if requiresEntitySlug, trimmedSlug.isEmpty {
            errorMessage = "Enter your organization slug (e.g. mosa-kuwait)."
            return
        }

        let slugForAPI = trimmedSlug.isEmpty ? nil : trimmedSlug

        do {
            let response = try await AuthAPI.login(
                username: trimmedUser,
                password: password,
                entitySlug: slugForAPI
            )
            guard let tokenValue = response.token else {
                errorMessage = "No token returned."
                return
            }
            try KeychainTokenStore.save(token: tokenValue)
            token = tokenValue
            user = response.user
            entities = response.entities ?? []
            password = ""
            PushRegistration.shared.bindAuthToken(tokenValue)
            await PushRegistration.shared.requestPermissionAndRegister()
            await loadHomeData()
            _ = await OfflineSubmitQueue.shared.flush(authToken: tokenValue)
        } catch let apiError as APIError {
            if apiError.code == "organization_required" {
                if let list = apiError.entities, !list.isEmpty {
                    directoryEntities = list
                }
                errorMessage = "Enter your organization slug, then your username."
            } else {
                errorMessage = apiError.message
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func refreshSession() async {
        guard token != nil else { return }
        isLoading = true
        defer { isLoading = false }
        let ok = await validateStoredSession()
        if !ok {
            errorMessage = "Session expired. Please sign in again."
        }
    }

    @discardableResult
    private func validateStoredSession() async -> Bool {
        guard let token else { return false }
        do {
            let response = try await AuthAPI.me(token: token)
            user = response.user
            entities = response.entities ?? []
            errorMessage = nil
            PushRegistration.shared.bindAuthToken(token)
            await PushRegistration.shared.requestPermissionAndRegister()
            await loadHomeData()
            _ = await OfflineSubmitQueue.shared.flush(authToken: token)
            return true
        } catch let apiError as APIError where apiError.code == "authentication_required" {
            handleSessionExpired()
            return false
        } catch {
            signOutLocal()
            return false
        }
    }

    func handleSessionExpired() {
        guard token != nil else { return }
        sessionExpiredMessage = "Your session expired. Please sign in again."
        signOutLocal()
    }

    func clearSessionExpiredMessage() {
        sessionExpiredMessage = nil
    }

    func loadHomeData() async {
        guard let token else { return }
        isLoadingHome = true
        defer { isLoadingHome = false }

        do {
            homeSummary = try await HomeAPI.fetchSummary(token: token)
        } catch {
            errorMessage = error.localizedDescription
        }

        do {
            publishedForms = try await HomeAPI.fetchPublishedForms(token: token)
        } catch {
            if errorMessage == nil {
                errorMessage = error.localizedDescription
            }
        }

        do {
            let inbox = try await HomeAPI.fetchInbox(token: token)
            inboxItems = inbox.items
            inboxTotal = inbox.total
        } catch {
            if errorMessage == nil {
                errorMessage = error.localizedDescription
            }
        }
    }

    func refreshInbox() async {
        guard let token else { return }
        isLoadingHome = true
        defer { isLoadingHome = false }
        do {
            async let summary = HomeAPI.fetchSummary(token: token)
            async let inbox = HomeAPI.fetchInbox(token: token)
            let (s, i) = try await (summary, inbox)
            homeSummary = s
            inboxItems = i.items
            inboxTotal = i.total
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func isInboxActing(_ submissionId: Int) -> Bool {
        inboxActingIds.contains(submissionId)
    }

    func performInboxWorkflow(
        submissionId: Int,
        decision: String,
        comment: String?,
        workflowActionAnchor: Int?
    ) async {
        guard let token else { return }
        inboxActingIds.insert(submissionId)
        inboxItemErrors.removeValue(forKey: submissionId)
        defer { inboxActingIds.remove(submissionId) }

        do {
            _ = try await HomeAPI.performWorkflow(
                token: token,
                submissionId: submissionId,
                decision: decision,
                comment: comment,
                workflowActionAnchor: workflowActionAnchor
            )
            await refreshInbox()
        } catch let apiError as APIError {
            inboxItemErrors[submissionId] = apiError.message
        } catch {
            inboxItemErrors[submissionId] = error.localizedDescription
        }
    }

    func logout() async {
        if let token {
            isLoading = true
            await PushRegistration.shared.unregister()
            _ = try? await AuthAPI.logout(token: token)
            isLoading = false
        }
        signOutLocal()
    }

    func signOutLocal() {
        PushRegistration.shared.bindAuthToken(nil)
        KeychainTokenStore.delete()
        token = nil
        user = nil
        entities = []
        password = ""
        homeSummary = nil
        publishedForms = []
        inboxItems = []
        inboxTotal = 0
        inboxActingIds = []
        inboxItemErrors = [:]
    }
}
